File: sample/src/main/java/ru/terrakok/cicerone/sample/ui/main/BaseFragment.java
package ru.terrakok.cicerone.sample.ui.main;

import android.content.Context;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import moxy.MvpAppCompatFragment;

import java.lang.ref.WeakReference;
import java.util.Iterator;
import java.util.List;

public abstract class BaseFragment extends MvpAppCompatFragment {

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        FragmentActivity activity = getActivity();
        if (activity instanceof ChainHolder) {
            ((ChainHolder) activity).getChain().add(new WeakReference<Fragment>(this));
        }
    }

    @Override
    public void onDetach() {
        FragmentActivity activity = getActivity();
        if (activity instanceof ChainHolder) {
            List<WeakReference<Fragment>> chain = ((ChainHolder) activity).getChain();
            for (Iterator<WeakReference<Fragment>> it = chain.iterator(); it.hasNext();) {
                WeakReference<Fragment> fragmentReference = it.next();
                Fragment fragment = fragmentReference.get();
                if (fragment != null && fragment == this) {
                    it.remove();
                    break;
                }
            }
        }
        super.onDetach();
    }
}

File: sample/src/main/java/com/github/terrakok/cicerone/sample/ui/main/BaseFragment.kt
package com.github.terrakok.cicerone.sample.ui.main

import android.content.Context
import androidx.fragment.app.Fragment
import moxy.MvpAppCompatFragment
import java.lang.ref.WeakReference

abstract class BaseFragment : MvpAppCompatFragment(), ChainScreen {
    override fun onAttach(context: Context) {
        super.onAttach(context)

        val activity = activity
        if (activity is ChainHolder) {
            (activity as ChainHolder).chain.add(WeakReference<Fragment>(this))
        }
    }

    override fun onDetach() {
        val activity = activity
        if (activity is ChainHolder) {
            val chain = (activity as ChainHolder).chain
            val it = chain.iterator()
            while (it.hasNext()) {
                val fragmentReference = it.next()
                val fragment = fragmentReference.get()
                if (fragment != null && fragment === this) {
                    it.remove()
                    break
                }
            }
        }
        super.onDetach()
    }
}